internal/provider/data_source_http.go

```diff
@@ -0,0 +1,312 @@
+			},
+
+			"request_headers": {
+				Description: "A map of request header field names and values.",
+				Type: types.MapType{
+					ElemType: types.StringType,
+				},
+				Optional: true,
+			},
+
+			"request_timeout": {
+				Description: "The request timeout in milliseconds.",
+				Type:        types.Int64Type,
+				Optional:    true,
+				Validators: []tfsdk.AttributeValidator{
+					int64validator.AtLeast(1),
+				},
+			},
+
+			"retry": {
+				Description: "Retry request configuration.",
+				Attributes: tfsdk.SingleNestedAttributes(map[string]tfsdk.Attribute{
+					"attempts": {
+						Description: "The number of retry attempts.",
+						Type:        types.Int64Type,
+						Optional:    true,
+						Validators: []tfsdk.AttributeValidator{
+							int64validator.AtLeast(0),
+						},
+					},
+				}),
+				Optional: true,
+			},
+
+			"response_body": {
+				Description: "The response body returned as a string.",
+				Type:        types.StringType,
+				Computed:    true,
+			},
+
+			"response_headers": {
+				Description: `A map of response header field names and values.` +
+					` Duplicate headers are concatenated according to [RFC2616](https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2).`,
+				Type: types.MapType{
+					ElemType: types.StringType,
+				},
+				Computed: true,
+			},
+
+			"status_code": {
+				Description: `The HTTP response status code.`,
+				Type:        types.Int64Type,
+				Computed:    true,
+			},
+
+			"id": {
+				Description: "The ID of this resource.",
+				Type:        types.StringType,
+				Computed:    true,
+			},
+		},
+	}, nil
+}
+
+func (d *httpDataSourceType) NewDataSource(context.Context, tfsdk.Provider) (tfsdk.DataSource, diag.Diagnostics) {
+	return &httpDataSource{}, nil
+}
+
+var _ tfsdk.DataSource = (*httpDataSource)(nil)
+
+type httpDataSource struct{}
+
+func (d *httpDataSource) Read(ctx context.Context, req tfsdk.ReadDataSourceRequest, resp *tfsdk.ReadDataSourceResponse) {
+	var model modelV0
+	diags := req.Config.Get(ctx, &model)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	var retry retryModel
+
+	if !model.Retry.Null && !model.Retry.Unknown {
+		diags = model.Retry.As(ctx, &retry, types.ObjectAsOptions{})
+		resp.Diagnostics.Append(diags...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+	}
+
+	url := model.URL.Value
+	headers := model.RequestHeaders
+	timeout := model.RequestTimeout
+
+	var cancel context.CancelFunc
+
+	if !timeout.IsNull() {
+		ctx, cancel = context.WithTimeout(ctx, time.Duration(timeout.Value)*time.Millisecond)
+		defer cancel()
+	}
+
+	retryClient := retryablehttp.NewClient()
+	retryClient.Logger = levelledLogger{ctx}
+	retryClient.RetryMax = int(retry.Attempts.Value)
+	retryClient.CheckRetry = func(ctx context.Context, resp *http.Response, err error) (bool, error) {
+		if err == nil {
+			return false, nil
+		}
+
+		return true, fmt.Errorf("retrying as request generated error: %w", err)
+	}
```

Comment:
To me this is a sane approach.
Given we made a choice of letting user deal with error codes themselves, it feels like an actual client error is the only case when retrying make sense.

Reply:
I have taken a second to look at how the `retryablehttp` client does it, and they have this:

```golang
	// Check the response code. We retry on 500-range responses to allow
	// the server time to recover, as 500's are typically not permanent
	// errors and may relate to outages on the server side. This will catch
	// invalid response codes as well, like 0 and 999.
	if resp.StatusCode == 0 || (resp.StatusCode >= 500 && resp.StatusCode != http.StatusNotImplemented) {
		return true, fmt.Errorf("unexpected HTTP status %s", resp.Status)
	}
```

While I agree that it feels rather odd to ignore error codes, while also checking them, here they are making a fair point: most of 5xx errors are usually temporary issues on the server side of things. Yes, `post-conditions` will allow TF execution to stop in case of error, but for this class of issues (temporary server-side issues), it would end up having to fail the overall TF execution, and then the user has to find a way to retry.

I'm frankly on the fence about this. Both side of this argument have merits.

Reply:
Yep. Agreed. I'm on the fence too.